receiver/dockerstatsreceiver/receiver.go

```diff
@@ -65,7 +58,16 @@ func (r *receiver) start(ctx context.Context, _ component.Host) error {
 		return err
 	}
 
-	r.client, err = docker.NewDockerClient(dConfig, r.settings.Logger)
+	// Restore the default dialer due to bug in v0.4.1 of go-connections/sockets.go
+	// which overrides the proto, and doesn't reset it when a new scheme is applied
+	var opts []dclient.Opt
+	if strings.HasPrefix(dConfig.Endpoint, "http") {
+		dialer := &net.Dialer{
+			Timeout: r.config.Timeout,
+		}
+		opts = append(opts, dclient.WithDialContext(dialer.DialContext))
+	}
+	r.client, err = docker.NewDockerClient(dConfig, r.settings.Logger, opts...)
```

Comment:
Would this affect the docker observer as well such that it'd be preferable to be in the internal `NewDockerClient()`?

Reply:
I would need to double check but it is fair to say it would.

Reply:
Yeah it's a good question. I think it would, but I'm not 100% convinced that the solution here is necessarily the desired general solution for the client. 

I just tried pinning `go-connections` to `0.4.0` and that seems to fix it. I wonder if that is a better solution here.

Reply:
Ok see https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/12322/commits/2709245f8aaa7b7d8906afe07e2dbab6c7a4f18a 

I've pinned to 0.4.0 and excluded 0.4.1. Potentially that should be done in the internal docker client package itself, but I think I will raise an issue for it and fix it separately from this PR. 

This is a more robust solution than checking if the endpoint starts with `http` and then overriding the dial context.